### PR TITLE
[QMS-313] UX improvements for POIs

### PR DESCRIPTION
**Pull request by pingurus**
_**Note: Do not delete any of the sections Answer them all. Replace the descriptive text by your answer**_

**What is the linked issue for this pull request (start with a `#`):**

QMS-#313

**Describe roughly what you have done:**

I added links in the waypoint description for some predefined osm tags.

**What steps have to be done to perform a simple smoke test:**

1. Open a poi file
2. Copy a POI over to the workspace
3. See that links, mails and phone numbers are clickable

**Does the code comply to the coding rules and naming conventions [Coding Guidelines](https://github.com/Maproom/qmapshack/wiki/DeveloperCodingGuideline):**

- [x] yes

**Is every user facing string in a tr() macro?**

- [x] yes

**Did you add the ticket number and title into the changelog? Keep the numeric order in each release block.**

- [x] yes, I didn't forget to change changelog.txt

### CODE REVIEW

**Comment by pingurus** [comment]
@kiozen @wthaem

Any more links you'd wish to add? OSM also supplies dedicated facebook/vk/instagram/whatsapp links, however with the code we have now, we cover the largest seven link sources: [tag history](https://taghistory.raifer.tech/#***/contact:instagram/&***/contact:website/&***/website/&***/url/&***/contact:facebook/&***/contact:vk/&***/contact:whatsapp/&***/contact:email/&***/contact:phone/&***/phone/&***/email/)
For the other ones someone would have to do some research how we can link to them.

**Comment by wthaem** [comment]
My guess is that the latest commit does already use proj8. If so, I can not yet compile this QMS commit/PR because I did not yet understand how to cope with libtiff required for the proj8 compilation.

**Comment by pingurus** [comment]
You're not on linux right? I documented a bit in #337 what I tried in order to build, but I doubt that's useful for you.

**Comment by wthaem** [comment]
@pingurus: Success! With kiozen's help I have now a way to compile and run QMS with proj version 8 under Windows 10 (there was one annoying QMS/proj issue: an unexpected format change in a projection string).

capture_001_16032021_201003

**Comment by pingurus** [comment]
@kiozen: I have another idea for this, so don't merge for now

**Comment by kiozen** [comment]
Imho this can be done better. This length list of OSM tags is just intimidating. Additionally the GPX waypoint definition has fields for some of the tags. 

Let me show. This is how a waypoint is created by the current code:
image

And this is how it might look like:

image

* The name must not be repeated
* All links a re in the link section and therefor available in the on-screen bubble
* The elevation is taken over in the elevation field.

Still the description field is kind of ugly. It would be nice to generate some natural text out of all tags but I guess this is quite complex. Are there other apps doing better?

**Comment by kiozen** [comment]
Btw it would be niche to have all added waypoints locked by default.

File: src/qmapshack/poi/CRawPoi.cpp [hunk 1]
```diff
@@ -74,8 +74,87 @@ poi_t CRawPoi::toPoi() const
     poi_t poi;
     poi.pos = coordinates;
     poi.name = getName();
-    poi.desc = data.join("<br>\n");
     poi.icon = garminIcon;
+
+    const QRegularExpression emailRegex ("(contact:email)=(.+)", QRegularExpression::UseUnicodePropertiesOption);
```
Comment (kiozen) [inline src/qmapshack/poi/CRawPoi.cpp:79]: Imho this should be ´static const` as you do not want to initialize the regex multiple times

File: src/qmapshack/poi/CRawPoi.cpp [hunk 2]
```diff
@@ -74,8 +74,87 @@ poi_t CRawPoi::toPoi() const
...
+    const QRegularExpression phoneRegex ("(contact:phone|phone|contact:mobile)=(.+)", QRegularExpression::UseUnicodePropertiesOption);
+    const QRegularExpression websiteRegex ("(contact:website|website|url)=(.+)", QRegularExpression::UseUnicodePropertiesOption);
+    const QRegularExpression osmRegex ("(osm_id)=(P|W|R)/(.+)", QRegularExpression::UseUnicodePropertiesOption);
+    const QRegularExpression wikipediaRegex ("(.*:)?(wikipedia)(:[A-z]{2})?=(.+)", QRegularExpression::UseUnicodePropertiesOption);
+    const QRegularExpression wikidataRegex ("(.*:)?(wikidata)=(.+)", QRegularExpression::UseUnicodePropertiesOption);
+    poi.desc = "";
+    bool firstLine = true;
+    for(QString line : data)
```
Comment (kiozen) [inline src/qmapshack/poi/CRawPoi.cpp:87]: `const QString& line` and an € in the constant reference bucket ;)

File: src/qmapshack/poi/CRawPoi.cpp [hunk 3]
```diff
@@ -74,8 +74,87 @@ poi_t CRawPoi::toPoi() const
...
+        }
+        else
+        {
+            poi.desc += "<br>\n";
+        }
+
+        QRegularExpressionMatch emailMatch = emailRegex.match(line);
+        if(emailMatch.hasMatch())
+        {
+            poi.desc += emailMatch.captured(1).trimmed() + "=<a href=\"mailto:" + emailMatch.captured(2).trimmed() + "\">" + emailMatch.captured(2).trimmed() + "</a>";
```
Comment (kiozen) [inline src/qmapshack/poi/CRawPoi.cpp:101]: if you take leading and trailing white space out of the regex capture you wouldn't need `trimmed()`.

File: src/qmapshack/poi/CRawPoi.cpp [hunk 4]
```diff
@@ -74,8 +74,87 @@ poi_t CRawPoi::toPoi() const
...
+        if(firstLine)
+        {
+            firstLine = false;
+        }
+        else
+        {
+            poi.desc += "<br>\n";
+        }
+
+        QRegularExpressionMatch emailMatch = emailRegex.match(line);
```
Comment (kiozen) [inline src/qmapshack/poi/CRawPoi.cpp:98]: `const QRegularExpressionMatch&` no need to create copies.

**Comment by pingurus** [comment]
What do you think of the UI now? 

Should we translate the links (i.e. website <=> Webseite)? And remove the leading `contact:` if it is there?

File: src/qmapshack/poi/CRawPoi.cpp (hunk 2 above)
Comment (pingurus) [reply src/qmapshack/poi/CRawPoi.cpp:90]: :moneybag: :wink:

File: src/qmapshack/gis/wpt/CGisItemWpt.h [hunk 5]
```diff
@@ -348,7 +348,7 @@ class CGisItemWpt : public IGisItem
     const searchValue_t getValueByKeyword(searchProperty_e keyword) override;
 
     static QString getLastName(const QString &name);
-    static void newWpt(const QPointF &pt, const QString& name, const QString& desc, IGisProject *project, const QString &icon = "", bool openEditWIndow = true);
+    static void newWpt(const QPointF &pt, const QString& name, const QString& desc, IGisProject *project, const QString &icon = "", const QList<link_t> links = QList<link_t>(), quint32 ele = NOINT, bool openEditWIndow = true);
```
Comment (kiozen) [inline src/qmapshack/gis/wpt/CGisItemWpt.h:351]: Instead of passing this endless list of parameters and default arguments wouldn't it be better to override newWpt and pass the POI object?

**Review by kiozen — CHANGES_REQUESTED** [review]
Yes translation if possible is always desired. 

Would it be possible to derive a nice address form all the address tags? Maybe it's better to store the tags not as a string but as a map. By that the regex voodoo isn't needed. And the conversion map to string could be used for the tooltip, too.

**Comment by pingurus** [comment]
> Yes translation if possible is always desired.
> 
> Would it be possible to derive a nice address form all the address tags? Maybe it's better to store the tags not as a string but as a map. By that the regex voodoo isn't needed. And the conversion map to string could be used for the tooltip, too.

Even with a map, the regex vodoo would still be needed (and more complicated), as for example for wikipedia I need the exact tag to form the URL. And to save resources, regexing should only happen upon conversion to a poi_t, so the map would only be created for `poi_t::toPoi`.

**Comment by pingurus** [comment]
In what format do you want the address to be? It might be a rather cumbersome task to generate valid address strings for each format used around the world :smile:

See for example https://wiki.openstreetmap.org/wiki/Proposed_features/AddrN

**Comment by pingurus** [comment]
@kiozen @wthaem Is the tooltip better now? A said above, I'm not sure how to handle addresses, but they should be transformed in some way, so both description and tooltip look better.

Regarding the code to generate the tooltip I think we can do some optimisation once we know how it should look.

**Comment by wthaem** [comment]
3 remarks:

* Replace "From category" by "Category"
* Replace "Has links to" by "Links" and suppress this line if there is no link
* Very nice trap: cope with left/right justified - see attachment. Waypoint and POI info should have the same justification and writing order (look especially at 8.30 - 11.0!) . And: if there is a link in the POI, then it is sometimes shown in the corresponding waypoint as proper link. But not always - see attachment. Why?

left_right

**Comment by kiozen** [comment]
> A said above, I'm not sure how to handle addresses, but they should be transformed in some way, so both description and tooltip look better.

Why not start simple with 
<street><house number>
<postcode><city>

If any of theses is missing, well it's missing. And all the other extensions are subject to further work if a good use-case can be presented.

**Comment by wthaem** [comment]
* address: maybe, add country, too?

* at least in my region phone numbers can not yet used as URL. Hiding them behind a link in the waypoint edit window is therefore not the best idea. Can the phone number be shown without a link? Windows10: if you click the phone link, you are asked which app to use to open the link.

**Comment by pingurus** [comment]
I've pushed an intermediate piece of code that is not really tested and for example breaks the assignment of the poi icons.

It would be nice if you'd still take a look at the tooltips and waypoints regarding the tags, so I can fix more bugs when I find time to code the next time :smile:

I've also switched to maps for the tags (yeah, I thought about that a bit and it makes life much easier for implementing filtering based on tags :wink:). The loading of the POIs may thus be a bit slower as more code runs in the constructor of the `CRawPoi` items, but I think it's not noticable.

**Comment by pingurus** [comment]
Note to self/note so you don't need to mention it: phone numbers are suppressed right now.

File: src/qmapshack/poi/CRawPoi.cpp [hunk 6]
```diff
@@ -19,63 +19,344 @@
...
+
+    return desc;
+}
+
+QList<IGisItem::link_t> CRawPoi::getLinks() const
+{
+    QList<IGisItem::link_t> links;
+    if(data.contains("contact:email"))
+    {
+        links += {"mailto:" + data["contact:email"], QObject::tr("Email"), "text/html"};
```
Comment (kiozen) [inline src/qmapshack/poi/CRawPoi.cpp:235]: Please no `QObject::tr()` Use `Q_DECLARE_TR_FUNCTIONS` macro to define a local `tr()`

Comment (pingurus) [reply src/qmapshack/poi/CRawPoi.cpp:235]: I knew there was some other way to do it :thinking:

Comment (kiozen) [reply src/qmapshack/poi/CRawPoi.cpp:235]: Yes and by that Linguist is listing the stuff as part of CRawPoi helping the translators to find the correct context.

**Comment by pingurus** [comment]
@wthaem @kiozen 
Any more remarks regarding the tooltip?

Is there anything I can do about the left/right alignment/writing direction of the tooltips or is this a bug somewhere deep in qt/QMS?

**Comment by kiozen** [comment]
Tool tips are looking fine for now. I guess we will tinker on the details in the future. But for now it's ok. However converting the POI to a waypoint the text is lost. Can you please have a look?

And you have to address the clang/clazy warnings. I added the topic to the coding guidelines lately. 

> Use Clang-Tidy and Clazy. All pull requests will be tested with these tools. They won't be merged on warnings unless it's definitely a false positive. QtCreator has support for these tools. Use Analyze->Clang-Tidy and Clazy to check all files you changes. Don't start a full scan as it takes hours.

**Comment by kiozen** [comment]
I found another "issue". If there is no address there are quite some empty lines between the category and the tags. This should be reduced to a single empty line.

**Comment by pingurus** [comment]
> However converting the POI to a waypoint the text is lost. Can you please have a look?

Can you provide an example? It works for me.

File: src/qmapshack/gis/wpt/CGisItemWpt.cpp [hunk 7]
```diff
@@ -285,6 +286,29 @@ void CGisItemWpt::newWpt(const QPointF& pt, const QString &name, const QString &
...
+    cfg.setValue("Waypoint/lastIcon", wpt->getIconName());
+}
+
+void CGisItemWpt::newWpt(const poi_t &poi, IGisProject *project, bool openEditWIndow)
+{
+    SETTINGS;
+    const QString& _icon = poi.icon.isEmpty() ? cfg.value("Waypoint/lastIcon", "Waypoint").toString() : poi.icon;
+    const QString& _name = poi.name.isEmpty() ? getLastName("") : poi.name;
+
+    CGisItemWpt * wpt = new CGisItemWpt(poi.pos * RAD_TO_DEG, _name, _icon, project);
```
Comment (kiozen) [inline src/qmapshack/gis/wpt/CGisItemWpt.cpp:299]: I would suggest to set the new waypoint initialy to read-only. By that the information in the waypoint's edit dialog is more compact. And most likely no one will edit the waypoint anyways.

`wpt->setReadOnlyMode(true);`

**Review by kiozen — CHANGES_REQUESTED** [review]

**Comment by kiozen** [comment]
> Can you provide an example? It works for me.

All good. Looks like I messed up.

**Comment by pingurus** [comment]
> Tool tips are looking fine for now. I guess we will tinker on the details in the future. But for now it's ok. ~ However converting the POI to a waypoint the text is lost. Can you please have a look?~ Forget about that. It's working. I probably started the wrong binary.
> 
> And you have to address the clang/clazy warnings. I added the topic to the coding guidelines lately.
> 
> > Use Clang-Tidy and Clazy. All pull requests will be tested with these tools. They won't be merged on warnings unless it's definitely a false positive. QtCreator has support for these tools. Use Analyze->Clang-Tidy and Clazy to check all files you changes. Don't start a full scan as it takes hours.

When I try to start Clang-Tidy and Clazy from qtcreator I get the following error: `10:11:04: Clazy: Invalid executable "". Stopped.`. Is there a quick fix for that that I'm not aware of?

**Comment by kiozen** [comment]
> > Tool tips are looking fine for now. I guess we will tinker on the details in the future. But for now it's ok. ~ However converting the POI to a waypoint the text is lost. Can you please have a look?~ Forget about that. It's working. I probably started the wrong binary.
> > And you have to address the clang/clazy warnings. I added the topic to the coding guidelines lately.
> > > Use Clang-Tidy and Clazy. All pull requests will be tested with these tools. They won't be merged on warnings unless it's definitely a false positive. QtCreator has support for these tools. Use Analyze->Clang-Tidy and Clazy to check all files you changes. Don't start a full scan as it takes hours.
> 
> When I try to start Clang-Tidy and Clazy from qtcreator I get the following error: `10:11:04: Clazy: Invalid executable "". Stopped.`. Is there a quick fix for that that I'm not aware of?

Usually this should work out of the box. Maybe your designer is outdated. I always install mine with the installation tool provided by Qt.

**Comment by pingurus** [comment]
Can I just more or less blindly accept the proposed fixits?

**Comment by pingurus** [comment]
Good Idea to include the tool, so you don't have to remind me of all the `const QString&` :laughing:

**Comment by kiozen** [comment]
> Can I just more or less blindly accept the proposed fixits?

No not really. In most cases they are ok, however there are many cases where the problem would have been solved if the provided container wouldn't be an instance but a const reference. e.g.

```
QStringList keys  = myMap.keys()

...

for(const QString& key : keys)
{
}
```

The automatic fix would simply wrap `keys` by `qAsConst(keys)`. However the proper fix would be to write 
```
const QStringList& keys  = myMap.keys()
```

avoiding the stupid copy of the list.

Therefore check all automatic fixes first and select only those you really want to apply. And fix the rest properly

> Good Idea to include the tool, so you don't have to remind me of all the const QString&

This was exactly the motivation to clean up the code to make the tool work in a sensible way.

**Comment by pingurus** [comment]
@kiozen So you prefer if we don't use the `qAsConst` even if the 'proper fix' would also just be to introduce a new throwaway variable that is `const`? Or is the code in this PR fine with using the `qAsConst`?

**Comment by kiozen** [comment]
As far as I understood the problem you have to use qAsConst for lvalues and a local const ref for rvalues. So using qAsConst is fine.

**Comment by pingurus** [comment]
I think we can merge this and I'll start a new PR for the filter.
Continuing on #352 only really makes sense if we have a stable interface, which is likely to change whilst implementing the filters.

**Comment by kiozen** [comment]
I still miss a push fixing all the clazy issues.

**Comment by pingurus** [comment]
Isn't  a779457 fixing all of them?

**Comment by kiozen** [comment]
@pingurus  Nope:

image

**Comment by pingurus** [comment]
Then I've missed out those files :thinking:

**Comment by pingurus** [comment]
How should I handle the last warning? (The one in CPoiPropSetup.cpp)

**Comment by pingurus** [comment]
I don't get why we get this warning here, as we don't call the virtual function in the base class but in the derived class and according to the [ISO CPP FAQ](https://isocpp.org/wiki/faq/strange-inheritance#calling-virtuals-from-ctors ) the problem arises when you call a virtual function in the ctor of the base class. Or did I get something wrong?

**Comment by kiozen** [comment]
Well, but you could derive from `CPoiPropSetup` and override `slotPropertiesChanged()`. In that case it's not working properly and therefore it's better to call `CPoiPropSetup::slotPropertiesChanged();` explicitly.

**Comment by kiozen** [comment]
I will give it another functional check and code review, later. And if nothing else pops up I would say we merge this into dev. Agreed or do you want to fix something else before hitting dev?

File: src/qmapshack/gis/wpt/CGisItemWpt.cpp [hunk 8]
```diff
@@ -285,6 +286,30 @@ void CGisItemWpt::newWpt(const QPointF& pt, const QString &name, const QString &
         wpt->setDescription(desc);
     }
```
Comment (kiozen) [inline src/qmapshack/gis/wpt/CGisItemWpt.cpp:288]: Why did you remove 

```
wpt->editInitial();
```

This is changing the expected behavior for adding waypoints.

File: src/qmapshack/poi/CPoiPOI.cpp [hunk 9]
```diff
@@ -318,16 +326,35 @@ bool CPoiPOI::getToolTip(const QPoint &px, QString &str) const
             {
                 str += "<b>" + name + "</b><br>\n";
             }
-            str += tr("From category ") + "<b>" + poiFound.getCategory() + "</b><br>\n";
-            str += poiFound.getData().join("<br>\n");
+            str += tr("Category: ") + "<b>" + poiFound.getCategory() + "</b><br>\n";
+            str += poiFound.getDesc();
+            str += "<br>\n";
+            const QList<IGisItem::link_t> links = poiFound.getLinks();
```
Comment (kiozen) [inline src/qmapshack/poi/CPoiPOI.cpp:332]: no reference?

**Review by kiozen — CHANGES_REQUESTED** [review]

File: src/qmapshack/gis/wpt/CGisItemWpt.cpp (hunk 8 above)
Comment (pingurus) [reply src/qmapshack/gis/wpt/CGisItemWpt.cpp:288]: I did not remove it, I moved into a `if` which defaults to true

Comment (kiozen) [reply src/qmapshack/gis/wpt/CGisItemWpt.cpp:288]: It's missing for 
```
void CGisItemWpt::newWpt(const QPointF& pt, const QString &name, const QString &desc, IGisProject * project)
```

Just compare QMS from `dev` and your branch. The one on dev opens the edit dialog on a new waypoint. Your branch simply sets the waypoint.

Comment (pingurus) [reply src/qmapshack/gis/wpt/CGisItemWpt.cpp:288]: Ok, my mistake. The way the side by side diff pulled it apart I didn't notice that it was only present in one of the methods :laughing:

Comment (kiozen) [reply src/qmapshack/gis/wpt/CGisItemWpt.cpp:288]: no worries, all good :D

**Comment by kiozen** [comment]
I hate to spoil the party but with latest commit I can not load pois anymore.

* The poi file activates
* If I select the first category in the list it loads and displays.
* If I select other categories they are not loaded and displayed.

**Comment by pingurus** [comment]
The suggestion of clazy broke it. Why? No idea :laughing: 
It seems that iterating over `foo bar: baz`is not necessarily the same as `foo bar: baz.keys()`even though clazy says so and I thought so aswell.

File: src/qmapshack/poi/CRawPoi.cpp [hunk 10]
```diff
@@ -19,63 +19,363 @@
...
+        }
+        desc += "<b>" + tr("Fax: ") + "</b>" + data["fax"];
+    }
+    if(!desc.isEmpty())
+    {
+        desc += "<br>\n" "<br>\n";
+    }
+    // Append all other tags
+    bool firstTag = true;
+    for(const QString& key : data)
```
Comment (kiozen) [inline src/qmapshack/poi/CRawPoi.cpp:216]: I think there is another bug. It should read 
```
    const QStringList& keys = data.keys();
    for(const QString& key : keys)
```
 Could you please check all other locations where you used range loops, too. And actually test you code before you commit ;)

**Review by kiozen — CHANGES_REQUESTED** [review]

File: src/qmapshack/poi/CRawPoi.cpp (hunk 10 above)
Comment (pingurus) [reply src/qmapshack/poi/CRawPoi.cpp]: Sorry. I tested it better now and found a segmentation fault when selecting POIs from the map but can't figure out it's cause.
It happens right after you selected whether you want to choose a symbol manually or not when you prompt it to copy the selected items over to a project, regardless of your response. It however either faults in the `SETTINGS` macro if you select `No` and in `QString::isEmpty()` if you select `yes`. The mutex locker is there, so that's not the cause and the pois that should be copied look intact. Do you have an idea what's the cause?

Comment (kiozen) [reply src/qmapshack/poi/CRawPoi.cpp]: start with gdb and analyze the backtrace?

Comment (kiozen) [reply src/qmapshack/poi/CRawPoi.cpp]: Btw we still have issus #357, too. Seems to be only on the POI feature branch. Probably some mutex problem.

Comment (kiozen) [reply src/qmapshack/poi/CRawPoi.cpp]: Just to keep track: The segfault you mentioned is not fixed, yet. Issue #357 is not reproducible. Correct?

Comment (kiozen) [reply src/qmapshack/poi/CRawPoi.cpp]: Bump! Any news on that?

Comment (pingurus) [reply src/qmapshack/poi/CRawPoi.cpp]: Didn't have the nerves yet to to investigate the reason, as I currently code much for work.
But I think I will spend some time on it this weekend, as it was a short working week :wink:

Comment (pingurus) [reply src/qmapshack/poi/CRawPoi.cpp]: > Just to keep track: The segfault you mentioned is not fixed, yet. Issue #357 is not reproducible. Correct?

Correct. 

Could you please check if you get the segfault when you try to select items on the map and copy them to a project? I think I have it traced down to the `exec()`calls of the dialogs that spin an additional event loop after which it faults when accessing the `poi_t`that it currently copies. Since it is only after the dialogs that it fails, you have to select an item that has the default POI symbol. Since it is something that definitely worked before, I'd like to check if it might have to do with my flawed libgdal setup (although I doubt it).

Comment (kiozen) [reply src/qmapshack/poi/CRawPoi.cpp]: I couldn't reproduce the crash. However I noticed another bug. You should move the line 

```
wpt->setReadOnlyMode(true);
```

behind `wpt->editInitial();`  else the icon is not editable (unless you unlock the item first)

I just wonder if the crash also occurs when you run QMS with gdb. Or does it produce a core dump that can be analyzed. Are there any Qt related messages when running with `-d`?

Comment (pingurus) [reply src/qmapshack/poi/CRawPoi.cpp]: > I couldn't reproduce the crash. However I noticed another bug. You should move the line
> 
> ```
> wpt->setReadOnlyMode(true);
> ```
> 
> behind `wpt->editInitial();` else the icon is not editable (unless you unlock the item first)

I'll do that :+1: 

> Are there any Qt related messages when running with -d?

There were, but that was related to another edge-case bug that occurred by chance which I managed to fix :laughing: 
I used `qDebug()` to trace down the issue to the `exec()` as `qDebug() << "icon: " << poi.icon;` works before doing `exec()` on the dialog window but not after (at least for me).

File: src/qmapshack/mouse/CMouseSelect.cpp [hunk 11]
```diff
@@ -26,6 +26,8 @@
 #include "mouse/CScrOptSelect.h"
 #include <QtWidgets>
 
+QMutex CMouseSelect::mutexPoisFound(QMutex::Recursive);
```
Comment (pingurus) [inline src/qmapshack/mouse/CMouseSelect.cpp:29]: @kiozen could you please have a look why this mutex isn't effective in stopping `draw()` being run parallel to `slotCopy()`? I'm pretty sure my segfaults are due to `poisFound` being updated while the copying is happening and can't figure out why this mutex doesn't lock.

Comment (kiozen) [reply src/qmapshack/mouse/CMouseSelect.cpp:29]: Because  `CMouseSelect::draw()` and `CMouseSelect::slotCopy()` are running in the same thread. See `void CCanvas::paintEvent(QPaintEvent*)` line 640. This is the main draw loop that resides in the main thread. The additional thread is triggered by `void IDrawContext::draw(QPainter& p, CCanvas::redraw_e needsRedraw, const QPointF& f)`.

Comment (kiozen) [reply src/qmapshack/mouse/CMouseSelect.cpp:29]: Maybe you get away with it by setting the `mutexPoisFound` to non-recursive and quit `findItems` when the mutex is locked. but I am not sure if that works in the same thread. Or you create a temporary real copy of the POI first. By that you can enter the eventloop and `findItems` can recreate the list without destroying your POI to copy.

**Comment by pingurus** [comment]
Your suggestion for the mutex worked, so I think this is done now. I still can't reproduce #357, so I can't fix that.
